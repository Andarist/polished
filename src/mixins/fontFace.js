// @flow
import validateModule, { customRule } from '../validation/_validateModule'

function generateFileReferences(
  fontFilePath: string,
  fileFormats: Array<string>,
) {
  const fileFontReferences = fileFormats.map(
    format => `url("${fontFilePath}.${format}")`,
  )
  return fileFontReferences.join(', ')
}

function generateLocalReferences(localFonts: Array<string>) {
  const localFontReferences = localFonts.map(font => `local("${font}")`)
  return localFontReferences.join(', ')
}

function generateSources(
  fontFilePath?: string,
  localFonts?: Array<string>,
  fileFormats: Array<string>,
) {
  const fontReferences = []
  if (localFonts) fontReferences.push(generateLocalReferences(localFonts))
  if (fontFilePath) {
    fontReferences.push(generateFileReferences(fontFilePath, fileFormats))
  }
  return fontReferences.join(', ')
}

/**
 * CSS for a @font-face declaration.
 *
 * @example
 * // Styles as object basic usage
 * const styles = {
 *    ...fontFace({
 *      'fontFamily': 'Sans-Pro'
 *      'fontFilePath': 'path/to/file'
 *    })
 * }
 *
 * // styled-components basic usage
 * injectGlobal`${
 *   fontFace({
 *     'fontFamily': 'Sans-Pro'
 *     'fontFilePath': 'path/to/file'
 *   }
 * )}`
 *
 * // CSS as JS Output
 *
 * '@font-face': {
 *   'fontFamily': 'Sans-Pro',
 *   'src': 'url("path/to/file.eot"), url("path/to/file.woff2"), url("path/to/file.woff"), url("path/to/file.ttf"), url("path/to/file.svg")',
 * }
 */

function fontFace(
  fontFamily: string,
  fontFilePath?: string,
  fontStretch?: string,
  fontStyle?: string,
  fontVariant?: string,
  fontWeight?: string,
  fileFormats?: Array<string> = ['eot', 'woff2', 'woff', 'ttf', 'svg'],
  localFonts?: Array<string>,
  unicodeRange?: string,
) {
  if (
    !customRule('mixins/fontFace', {
      enforce: fontFilePath || localFonts,
      msg: 'fontFace expects either the path to the font file(s) or a name of a local copy. However, you provided neither.',
    })
  ) {
    return {}
  }

  const fontFaceDeclaration = {
    '@font-face': {
      fontFamily,
      src: generateSources(fontFilePath, localFonts, fileFormats),
      unicodeRange,
      fontStretch,
      fontStyle,
      fontVariant,
      fontWeight,
    },
  }

  // Removes undefined fields for cleaner css object.
  return JSON.parse(JSON.stringify(fontFaceDeclaration))
}
export default validateModule({
  modulePath: 'mixins/fontFace',
  arrityCheck: { min: 1, max: 9 },
  typeCheck: [
    {
      key: 'fontFamily',
      type: 'string',
      required: true,
    },
    { key: 'fontFilePath', type: 'string' },
    { key: 'fontStretch', type: 'string' },
    { key: 'fontStyle', type: 'string' },
    { key: 'fontVariant', type: 'string' },
    { key: 'fontWeight', type: 'string' },
    { key: 'fileFormats', type: 'array' },
    { key: 'localFonts', type: 'array' },
    { key: 'unicodeRange', type: 'string' },
  ],
})(fontFace)
