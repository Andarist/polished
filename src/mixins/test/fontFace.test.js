// @flow
import fontFace from '../fontFace'

describe('fontFace', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should return a valid object when passed just a family and source', () => {
    expect({
      ...fontFace({
        fontFamily: 'Sans Pro',
        fontFilePath: 'path/to/file',
      }),
    }).toMatchSnapshot()
  })

  it('should return a valid object when passed both local and file-based sources', () => {
    expect({
      ...fontFace({
        fontFamily: 'Sans Pro',
        localFonts: ['sans-pro'],
        fontFilePath: 'path/to/file',
      }),
    }).toMatchSnapshot()
  })

  it('should return a valid object when passed both a file-based source and multiple local sources', () => {
    expect({
      ...fontFace({
        fontFamily: 'Sans Pro',
        localFonts: ['sans-pro', 'sans pro'],
        fontFilePath: 'path/to/file',
      }),
    }).toMatchSnapshot()
  })

  it('should return a valid object when passed only local sources', () => {
    expect({
      ...fontFace({
        fontFamily: 'Sans Pro',
        localFonts: ['sans-pro', 'sans pro'],
      }),
    }).toMatchSnapshot()
  })

  it('should respect the file format configuration object', () => {
    expect({
      ...fontFace({
        fontFamily: 'Sans Pro',
        fontFilePath: 'path/to/file',
        fileFormats: ['eot', 'svg'],
      }),
    }).toMatchSnapshot()
  })

  it('should set passed font configuration variables', () => {
    expect({
      ...fontFace({
        fontFamily: 'Sans Pro',
        fontFilePath: 'path/to/file',
        fontStretch: 'condensed',
        fontStyle: 'italic',
        fontWeight: 'bold',
        fontVariant: 'small-caps',
        unicodeRange: 'U+26',
      }),
    }).toMatchSnapshot()
  })

  it('should throw an error when not passed a fontFilePath or localFonts', () => {
    fontFace({
      fontFamily: 'Sans Pro',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not provided any parameters', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    fontFace()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when not passed a value for fontFamily', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    fontFace({
      fontFilePath: 'path/to/file',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for fontFamily', () => {
    fontFace({
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      fontFamily: 100,
      fontFilePath: 'path/to/file',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for fontFilePath', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      fontFilePath: 100,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for fontStyle', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      fontFilePath: 'path/to/file',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      fontStyle: 100,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for fontStretch', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      fontFilePath: 'path/to/file',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      fontStretch: true,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for fontWeight', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      fontFilePath: 'path/to/file',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      fontWeight: true,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for fontVariant', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      fontFilePath: 'path/to/file',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      fontVariant: 100,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-array value for localFonts', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      localFonts: 'sans-pro',
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when passed a non-string value for unicodeRange', () => {
    fontFace({
      fontFamily: 'Sans Pro',
      fontFilePath: 'path/to/file',
      // $FlowIgnoreNextLine since the coming is invalid code, flow complains
      unicodeRange: 26,
    })
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
})
