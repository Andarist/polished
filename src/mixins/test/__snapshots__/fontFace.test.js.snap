exports[`fontFace should respect the file format configuration object 1`] = `
Object {
  "@font-face": Object {
    "fontFamily": "Sans Pro",
    "src": "url(\"path/to/file.eot\"), url(\"path/to/file.svg\")",
  },
}
`;

exports[`fontFace should return a valid object when passed both a file-based source and multiple local sources 1`] = `
Object {
  "@font-face": Object {
    "fontFamily": "Sans Pro",
    "src": "local(\"sans-pro\"), local(\"sans pro\"), url(\"path/to/file.eot\"), url(\"path/to/file.woff2\"), url(\"path/to/file.woff\"), url(\"path/to/file.ttf\"), url(\"path/to/file.svg\")",
  },
}
`;

exports[`fontFace should return a valid object when passed both local and file-based sources 1`] = `
Object {
  "@font-face": Object {
    "fontFamily": "Sans Pro",
    "src": "local(\"sans-pro\"), url(\"path/to/file.eot\"), url(\"path/to/file.woff2\"), url(\"path/to/file.woff\"), url(\"path/to/file.ttf\"), url(\"path/to/file.svg\")",
  },
}
`;

exports[`fontFace should return a valid object when passed just a family and source 1`] = `
Object {
  "@font-face": Object {
    "fontFamily": "Sans Pro",
    "src": "url(\"path/to/file.eot\"), url(\"path/to/file.woff2\"), url(\"path/to/file.woff\"), url(\"path/to/file.ttf\"), url(\"path/to/file.svg\")",
  },
}
`;

exports[`fontFace should return a valid object when passed only local sources 1`] = `
Object {
  "@font-face": Object {
    "fontFamily": "Sans Pro",
    "src": "local(\"sans-pro\"), local(\"sans pro\")",
  },
}
`;

exports[`fontFace should set passed font configuration variables 1`] = `
Object {
  "@font-face": Object {
    "fontFamily": "Sans Pro",
    "fontStretch": "condensed",
    "fontStyle": "italic",
    "fontVariant": "small-caps",
    "fontWeight": "bold",
    "src": "url(\"path/to/file.eot\"), url(\"path/to/file.woff2\"), url(\"path/to/file.woff\"), url(\"path/to/file.ttf\"), url(\"path/to/file.svg\")",
    "unicodeRange": "U+26",
  },
}
`;

exports[`fontFace should throw an error when not passed a fontFilePath or localFonts 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cfontFace expects either the path to the font file(s) or a name of a local copy. However, you provided neither.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when not passed a value for fontFamily 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 1st parameter of type string. However, you did not pass one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when not provided any parameters 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a minimum of %c1 parameter%c. However, you passed %c0 parameters%c.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: green",
    "color: black; font-size: 12px",
    "color: black; font-size: 12px; font-weight: bold; color: red",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 1st parameter of type string. However, you did not pass one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-array value for localFonts 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 8th parameter of type array. However, you passed sans-pro(string) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for fontFamily 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 1st parameter of type string. However, you passed 100(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for fontFilePath 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 2nd parameter of type string. However, you passed 100(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for fontStretch 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 3rd parameter of type string. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for fontStyle 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 4th parameter of type string. However, you passed 100(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for fontVariant 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 5th parameter of type string. However, you passed 100(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for fontWeight 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 6th parameter of type string. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`fontFace should throw an error when passed a non-string value for unicodeRange 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/fontFace.js --%c

fontFace %cexpects a 9th parameter of type string. However, you passed 26(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#fontFace%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;
